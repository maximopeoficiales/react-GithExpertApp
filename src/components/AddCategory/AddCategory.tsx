import React, { SyntheticEvent, useState } from "react";
import PropTypes from "prop-types";

interface MyProps {
  addCategory(category: string): void;
}
const defaultProps = {
  // optionalText: "Soy un subtitulo opcional",
};

const AddCategory = (props: MyProps) => {
  props = { ...defaultProps, ...props };

  const [value, setValue] = useState("");
  const handlerChange = (e: any) => {
    setValue(e.target.value);
  };
  const handlerSubmit = (e: SyntheticEvent) => {
    e.preventDefault();
    if (value.trim().length > 2) {
      props.addCategory(value);
      setValue("");
    }
  };
  return (
    <div className="container-input">
      {/* <p>Search:</p> */}
      <form onSubmit={handlerSubmit}>
        <input
          type="text"
          value={value}
          className="input-query"
          onChange={handlerChange}
          placeholder="Search"
        />
      </form>
    </div>
  );
};

AddCategory.propTypes = {
  addCategory: PropTypes.func.isRequired,
};
export default AddCategory;
