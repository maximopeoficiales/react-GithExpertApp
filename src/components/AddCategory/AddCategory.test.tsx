import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import AddCategory from "./AddCategory";
import { shallow, ShallowWrapper } from "enzyme";

describe("<AddCategory />", () => {
  // al instanciarlo asi, me permite contar cuantas veces fue ejecutado
  const addCategory = jest.fn();
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    // borrar todas interacciones realizadas
    jest.clearAllMocks();
    // vuelve a  cargar el componente
    wrapper = shallow(<AddCategory addCategory={addCategory} />);
  });

  test("debe mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("debe cambiar la caja de texto", () => {
    let input = wrapper.find("input");
    const value = "HOla bro cambie el texto";
    // simula el evento y envia el evento como objecto
    input.simulate("change", { target: { value } });

    // se busca el input y se verifica si cambio su valor
    input = wrapper.find("input");
    const textValor = input.prop("value");
    console.log(textValor);

    expect(textValor).toBe(value);
  });

  test("No debe poster informacion con submit", () => {
    // simula el evento submit
    wrapper.find("form").simulate("submit", { preventDefault() {} });
    // se espera que la funcion no se halla llamado
    expect(addCategory).not.toHaveBeenCalled();
  });
  test("debe de llamar el addCategory y limpiar la caja de texto", () => {
    // simular inputChange
    let input = wrapper.find("input");
    const value = "HOla bro cambie el texto";
    // simula el evento y envia el evento como objecto
    input.simulate("change", { target: { value } });

    // simular evento submit
    wrapper.find("form").simulate("submit", { preventDefault() {} });
    console.log(wrapper.find("input").props());

    expect(addCategory).toHaveBeenCalled();
    // espera que se llamo 1 vez
    expect(addCategory).toHaveBeenCalledTimes(1);
    // espera que la funciona se llame con una funcion
    // expect(addCategory).toHaveBeenCalledWith(expect.any(Function));
    expect(addCategory).toHaveBeenCalledWith(expect.any(String));

    expect(wrapper.find("input").prop("value")).toBe("");
  });
});
