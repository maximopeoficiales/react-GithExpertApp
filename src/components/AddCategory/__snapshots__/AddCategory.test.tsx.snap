// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddCategory /> debe mostrarse correctamente 1`] = `
<div
  className="container-input"
>
  <form
    onSubmit={[Function]}
  >
    <input
      className="input-query"
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value=""
    />
  </form>
</div>
`;
